packages/core/src/storage/namespace.ts

```diff
@@ -0,0 +1,280 @@
+  return context.contractDef.id === referencedId
+    ? context.contractDef
+    : context.deref(['ContractDefinition'], referencedId);
+}
+
+/**
+ * Add namespaces and source locations for the given compilation context's contract.
+ * Does not include inherited contracts.
+ */
+function addContractNamespacesWithSrcs(
+  namespaces: Record<string, NamespaceWithSrcs>,
+  decodeSrc: SrcDecoder,
+  layout: StorageLayout,
+  context: CompilationContext,
+  origContractDef: ContractDefinition,
+) {
+  for (const node of context.contractDef.nodes) {
+    if (isNodeType('StructDefinition', node)) {
+      const storageLocation = getStorageLocationAnnotation(node);
+      if (storageLocation !== undefined) {
+        const origSrc = decodeSrc(getOriginalStruct(node.canonicalName, origContractDef));
+
+        if (namespaces[storageLocation] !== undefined) {
+          namespaces[storageLocation].srcs.push(origSrc);
+        } else {
+          namespaces[storageLocation] = {
+            namespace: getNamespacedStorageItems(node, decodeSrc, layout, context, origContractDef),
+            srcs: [origSrc],
+          };
+        }
+      }
+    }
+  }
+}
+
+/**
+ * Gets the storage location string from the `@custom:storage-location` annotation.
+ *
+ * For example, when using ERC-7201 (https://eips.ethereum.org/EIPS/eip-7201), the result will be `erc7201:<NAMESPACE_ID>`
+ *
+ * @param node The node that may have a `@custom:storage-location` annotation.
+ * @returns The storage location string, or undefined if the node does not have a `@custom:storage-location` annotation.
+ * @throws Error if the node has the annotation `@custom:storage-location` but it does not have exactly one argument.
+ */
+export function getStorageLocationAnnotation(node: Node) {
+  const doc = getDocumentation(node);
+  if (hasAnnotationTag(doc, 'storage-location')) {
+    const storageLocationArgs = getAnnotationArgs(doc, 'storage-location');
+    if (storageLocationArgs.length !== 1) {
+      throw new Error('@custom:storage-location annotation must have exactly one argument');
+    }
+    return storageLocationArgs[0];
+  }
+}
+
+/**
+ * Gets the storage items for the given struct node.
+ * Includes loading recursive type information, and adds slot and offset if they are available in the given compilation context's layout.
+ */
+function getNamespacedStorageItems(
+  node: StructDefinition,
+  decodeSrc: SrcDecoder,
+  layout: StorageLayout,
+  context: CompilationContext,
+  origContractDef: ContractDefinition,
+): StorageItem[] {
+  const astTypeMembers = getTypeMembers(node, { typeName: true });
+
+  const storageItems: StorageItem[] = [];
+
+  for (const astMember of astTypeMembers) {
+    let item: StorageItem = {
+      contract: context.contractDef.name,
+      label: astMember.label,
+      type: astMember.type,
+      src: decodeSrc({ src: getOriginalMemberSrc(node.canonicalName, astMember.label, origContractDef) }),
```

Comment:
What would be the benefit of storing the `astId`?  Especially since once the manifest is written, the manifest contents may no longer have any relation to solc output (since the solc output in the user's project may get deleted).

Reply:
> Especially since once the manifest is written, the manifest contents may no longer have any relation to solc output

This is a good point. No need to store it.